src/flag_gems/ops/exponential_.py

```
@@ -0,0 +1,121 @@
+        d1 = uint_to_uniform_float(paste_u64(r1, r3))
+        y0 = transform_exponential(d0, lambd, eps)
+        y1 = transform_exponential(d1, lambd, eps)
+        UNROLL = 2
+        start = tl.program_id(0).to(tl.uint64) * BLOCK * UNROLL
+        off_0 = start + tl.arange(0, BLOCK)
+        off_1 = off_0 + BLOCK
+        tl.store(out_ptr + off_0, y0, mask=off_0 < N, eviction_policy="evict_first")
+        tl.store(out_ptr + off_1, y1, mask=off_1 < N, eviction_policy="evict_first")
+    else:
+        f0 = uint_to_uniform_float(r0)
+        f1 = uint_to_uniform_float(r1)
+        f2 = uint_to_uniform_float(r2)
+        f3 = uint_to_uniform_float(r3)
+        y0 = transform_exponential(f0, lambd, eps)
+        y1 = transform_exponential(f1, lambd, eps)
+        y2 = transform_exponential(f2, lambd, eps)
+        y3 = transform_exponential(f3, lambd, eps)
+        UNROLL = 4
+        start = tl.program_id(0).to(tl.uint64) * BLOCK * UNROLL
+        off_0 = start + tl.arange(0, BLOCK)
+        off_1 = off_0 + BLOCK
+        off_2 = off_1 + BLOCK
+        off_3 = off_2 + BLOCK
+        tl.store(out_ptr + off_0, y0, mask=off_0 < N, eviction_policy="evict_first")
+        tl.store(out_ptr + off_1, y1, mask=off_1 < N, eviction_policy="evict_first")
+        tl.store(out_ptr + off_2, y2, mask=off_2 < N, eviction_policy="evict_first")
+        tl.store(out_ptr + off_3, y3, mask=off_3 < N, eviction_policy="evict_first")
+
+
+@triton.jit
+def paste_u64(hi: tl.uint32, lo: tl.uint32):
+    hi = hi.to(tl.uint64) << 32
+    x = hi | lo.to(tl.uint64)
+    return x
+
+
+@triton.jit
+def transform_exponential(u, lambd, eps):
+    eps1 = -0.5 * eps
+    is_min = u >= 1.0 + eps1
+    log = tl.where(is_min, eps1, tl.math.log(u))
+    v = -1.0 / lambd * log
+    return v
+
+
+def exponential_(x, lambd: float = 1.0, *, gen=None):
+    logging.debug("GEMS EXPONENTIAL_")
+    dtype = x.dtype
+    device = x.device
+    inplace = x.is_contiguous()
```

Performing inplace operation on a tensor with internal overlapping should raise a Runtime Exception.

---

Now it would raise a runtime error when copying data back.

```python
import torch
import flag_gems
flag_gems.enable()
x = torch.ones(2, device="cuda").broadcast_to(x, (3, 2))
x.exponential_()
```

---

I'll fix it.

---

Pytorch throws with exactly the same error. We'll just keep the current way.